### PR TITLE
Add Categories component to plugins browser

### PR DESCRIPTION
**Pull request by lsl**
#### Changes proposed in this Pull Request

* Adds a new discover section (component named `Categories`)

#### Testing instructions

* http://calypso.localhost:3000/plugins
* Should see new discover section
* Should be able to click on discover dropdown
* Should be able to select categories

Related to https://github.com/Automattic/wp-calypso/issues/62264

### CODE REVIEW

**Comment by lsl** [comment]
This PR is probably a little big in scope.

- Can drop the category cards from the first pass as I don't think we have illustrations/icons chosen yet. (@SaxonF / @vinimotaa)
- Can just run searches for these categories as search terms and do the category specific queries in a follow up PR

For this PR, lets just ship the dropdown category selector in the first pass and figure out category searching in a follow up.

**Comment by allilevine** [comment]
> Can drop the category cards from the first pass as I don't think we have illustrations/icons chosen yet. (@SaxonF / @vinimotaa)

@lsl Can you clarify the design need here? Are you asking if we can merge the dropdown, then add the category cards?

> Can just run searches for these categories as search terms and do the category specific queries in a follow up PR

Do you mean searching within categories, or how we'll display results for each category selected?

**Comment by lsl** [comment]
> @lsl Can you clarify the design need here? Are you asking if we can merge the dropdown, then add the category cards?

I should have been more clear with those pings, was just reminding them we need icon selection for each category before we can implement that part of the issue.

I was assuming it's ok to merge the dropdown and category cards separately and separate again from figuring out how the backend searches are going to work.

> Do you mean searching within categories, or how we'll display results for each category selected?

I mean when you click the "Widgets" category it just runs a search for "widgets" and displays those results.

We need to figure out category search in https://github.com/Automattic/wp-calypso/issues/62497 before we can really implement much of the calypso side searching calls so it makes sense to just avoid that part of the issue for now and focus on the design.

In a couple of days when a category search/filter endpoint is implemented we can use that instead. (we can probably open up a PR to just change the .org side of things straight away)

**Comment by Rolilink** [comment]
>  mean when you click the "Widgets" category it just runs a search for "widgets" and displays those results.

Sounds like a good fallback if we are not able to finish the categories search in time, my worry would be deploying something that we will remove a couple of days laters, maybe deploy this behind a feature flag and remove the feature flag at the end of the sprint with either categories search or with this behavior @lsl?

File: client/my-sites/plugins/categories/README.md [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# Categories
+
+This component is used to render the discover section of the plugin marketplace.
+
+## How to use
+
+```jsx
+import Discover from 'calypso/my-sites/plugins/discover';
```
Comment (Rolilink) [inline client/my-sites/plugins/categories/README.md]: nitpick: shouldn't this be named `Categories` and the path be `'calypso/my-sites/plugins/categories'`

Comment (lsl) [reply client/my-sites/plugins/categories/README.md]: Yes, was going to call it discover but renamed it to be more obvious

```suggestion
import Discover from 'calypso/my-sites/plugins/categories';
```

**Comment by lsl** [comment]
> Sounds like a good fallback if we are not able to finish the categories search in time, my worry would be deploying something that we will remove a couple of days laters, maybe deploy this behind a feature flag and remove the feature flag at the end of the sprint with either categories search or with this behavior @lsl?

Won't be removing anything just changing the search URL it hits to be `tag=` instead of `s=`.

edit: 

I think you might be right, this ties in with the search feature, we need to limit searches to this category after you've changed it, that means we probably need to pass the category into the doSearch function which currently isn't very exposed.

Next steps, I think we should move current search into a hook so we can access / modify it from anywhere in the plugin browser stack, should help clean up a lot of this prop drilling too.

**Comment by Rolilink** [comment]
> I think you might be right, this ties in with the search feature, we need to limit searches to this category after you've changed it, that means we probably need to pass the category into the doSearch function which currently isn't very exposed.

@lupus2k recently discovered that the dotorg search supports multi-tag, but the marketplace one don't. This should at least help us to fetch categorized dotorg plugins

File: client/my-sites/plugins/plugins-browser/index.jsx [hunk 2]
```diff
@@ -324,8 +327,9 @@ const PluginsBrowser = ( {
 				title={ translate( 'Plugins you need to get your projects done' ) }
 				searchTerms={ [ 'seo', 'pay', 'booking', 'ecommerce', 'newsletter' ] }
 			/>
-			<Categories onSelect={ ( c ) => doSearch( c.tags?.[ 0 ] || '' ) } />
+			<Categories onSelect={ ( c ) => doSearch( search, { tag: c.tags?.join( ',' ) || '' } ) } />
```
Comment (lsl) [inline client/my-sites/plugins/plugins-browser/index.jsx]: I'm thinking this shouldn't be using doSearch, instead we should be linking to the `/plugins/:category` route in the dropdown, if a search term is present we can link to `/plugins/:category?s=:search` instead.

This `c.tags?.join( ',' )` can be done in the searching logic directly by looking up the tags in the popularCategories map. (I'm planning to move popularCategories to a `useMarketplaceCategories` hook)

**Comment by lupus2k** [comment]
I removed the categories component when searching & remove the `Search results for` text when searching.

URL should change as @lsl said ( p1649928434978089-slack-C029JEQRVRT ) but there are some things we have to take care about:

- [ ] Remove tag from URL when searching
- [ ] Show the correct category when the URL has `tag` instead of discovery
- [ ] If URL has tag, `wpcom` don't search for it on reload

**Comment by lupus2k** [comment]
Dropdown already has CSS but not `Discover` section yet.

**Comment by lupus2k** [comment]
We made some changes with @Rolilink.
- Updated the categories URL with `/plugins/:category`.
- Moved categories list to separate file.
- Removed search box on category filter.
- Show correct category when filtered & reload page.

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7229209

Hi @lsl, could you please edit the description of this PR and add a screenshot for our translators? Ideally it'd include all of the following strings:
- `Customer Service`
- `Search Optimization`
- `Photo & Video`
- `Posts & Posting`

Thank you in advance!

**Comment by lsl** [comment]
Thanks for the commits on this, it got really close to something working (really nice to see category search working!) With the todo's, failing tests, and a few regressions still needing attention I felt like the scope was getting way out of control for one PR.

To move things forward I decided to scratch out a new PR just to handle the category switcher: https://github.com/Automattic/wp-calypso/pull/62873

I basically just took this PR (https://github.com/Automattic/wp-calypso/pull/62685), reset to trunk and stripped out anything not to do with category switching. From there I cleaned up some of the user interactions with search / breadcrumbs. I also had to move the categories list to a hook to do some filtering based on site type.

I did this in a new PR because I didn't want to blow away any of the tag searching work in this PR as from what I could tell most of it can be used pretty much as is. 

Once https://github.com/Automattic/wp-calypso/pull/62873 is merged it should be relatively easy to add all the categories as "allowed" and merge in the supporting code to handle searching by tag behind the scenes.

Update: wanted to see just how easy it would be to merge the search stuff in after, turns out it pretty easy :smile: https://github.com/Automattic/wp-calypso/pull/62881

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7229209).

**Pull request closed without merging**